src/dataflow/render/mod.rs

```diff
@@ -314,6 +319,61 @@ pub(crate) fn build_dataflow<A: Allocate>(
                             Envelope::Upsert(_) => unreachable!(),
                         };
 
+                        // Implement source filtering and projection.
+                        // At the moment this is strictly optional, but we perform it anyhow
+                        // to demonstrate the intended use.
+                        if let Some(operators) = src.operators.clone() {
+                            // Determine replacement values for unused columns.
+                            let source_type = src.desc.typ();
+                            let position_or = (0..source_type.arity())
+                                .map(|col| {
+                                    if operators.projection.contains(&col) {
+                                        Ok(col)
+                                    } else {
+                                        Err({
+                                            let typ = &source_type.column_types[col];
+                                            if typ.nullable {
+                                                Datum::Null
+                                            } else {
+                                                typ.scalar_type.dummy_datum()
```

Comment:
Can we replace this with Datum::Null? Or is something preventing us from doing that?

Reply:
I remain concerned about the contract surrounding who will actually look at these values. I think we want to ultimately ensure that no one does, and there is probably a bug if someone does, but I was being conservative here. I'll add a TODO to simplify and minimize.